Network/HTTP2/H2/Manager.hs

```diff
@@ -122,5 +124,29 @@ withTimeout (Manager timmgr var) action =
     T.withHandleKillThread timmgr (return ()) $ \th -> do
         tid <- myThreadId
         -- overriding ThreadWithoutTimeout
-        atomically $ modifyTVar var $ Map.insert tid $ ThreadWithTimeout th
+        atomically $ modifyManagedThreads_ var $ Map.insert tid $ ThreadWithTimeout th
         action th
+
+----------------------------------------------------------------
+
+newManagedThreads :: IO ManagedThreads
+newManagedThreads = WrapManagedThreads <$> newTVarIO Map.empty
+
+getManagedThreads :: ManagedThreads -> STM (Map ThreadId TimeoutHandle)
+getManagedThreads = readTVar . unwrapManagedThreads
+
+modifyManagedThreads
+    :: ManagedThreads
+    -> (Map ThreadId TimeoutHandle -> (Map ThreadId TimeoutHandle, a))
+    -> STM a
+modifyManagedThreads (WrapManagedThreads var) f = do
+    threads <- readTVar var
+    let (threads', result) = f threads
+    writeTVar var $! threads' -- strict update
```

Comment:
This is the critical line.

Reply:
I'm confused.
If this is critical, `Data.Map.Strict` is good enough.

Reply:
No, that is not correct. Without the `$!`, it doesn't matter which function we use, because it is never called! It will just be a thunk build up of `insert .. delete .. insert .. delete .. insert ..`. That's why it's critical to force the map to WHNF (not NF), so that we actually call the function we use.

Reply:
We are using `Strict` and `StrictData` but `threads'` is in a tuple.
That's why?

Reply:
The `improve-manager` branch now use `modifyTVar'`.

Reply:
No, that is not the reason. `Strict` does not make function application strict, it only makes bindings strict. So if we have a module in which `Strict` is _not_ enabled (standing in for `writeTVar`)

```haskell
module A where

ignore :: a -> String
ignore _ = "Hi"
```

then in

```haskell
{-# LANGUAGE Strict, StrictData #-}

module B where

import A

ex1, ex2, ex3 :: IO ()
ex1 = putStrLn (ignore undefined)
ex2 = putStrLn (ignore $! undefined)
ex3 = let x = undefined in putStrLn (ignore x)
```

`ex1` prints "Hi" whereas `ex2` and `ex3` both throw an exception.